activesupport/lib/active_support/core_ext/range/intersection.rb

```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+class Range
+  # The maximum sized range covered by both self AND other
+  # Returns nil if there is no overlap between self and other
+  def intersection(other)
+    unless other.is_a?(::Range)
+      raise TypeError, "Can't intersect a range by a(n) #{other.class}"
+    end
+
+    if (inverted_range = [self, other].find { |r| r.inverted? })
+      raise ArgumentError, "Intersection of inverted range (#{inverted_range}) is undefined"
+    end
+
+    if overlaps?(other)
```

Comment:
you might be interested in [this](https://github.com/hernanat/extended_range/blob/main/lib/core_extensions/range/checks.rb#L17) definition of `#overlaps?` which avoids extra object allocations